File: src/Python/Frame.py
import os
pid = os.getpid()
import pyjit

@pyjit.object(
    lang="cpp",
    Debug=True,
    LibName="self",
    Serialization =True,
    include_dir=["Data/src","Data/inc","Common"],
    ClassName="PyDataFrame",
    include=["PyDataFrame.h"],
    impl=["Data/inc/*.cpp","Data/src/*.cpp","Common/*.cpp"]
    ) 
class DataFrame:
    version:int="__bind__:head->version"
    Type:int ="__bind__:head->type"
    StartTime:int="__bind__:head->startTime"
    SourceId:str
    SrcAddr:str
    DstAddr:str
    Format0:int="__bind__:head->format[0]"
    Format1:int="__bind__:head->format[1]"
    Format2:int="__bind__:head->format[2]"
    Format3:int="__bind__:head->format[3]"
    Format4:int="__bind__:head->format[4]"
    Format5:int="__bind__:head->format[5]"
    Format6:int="__bind__:head->format[6]"
    Format7:int="__bind__:head->format[7]"
    Metadata0:int="__bind__:head->metadata[0]"
    Metadata1:int="__bind__:head->metadata[1]"
    Metadata2:int="__bind__:head->metadata[2]"
    Metadata3:int="__bind__:head->metadata[3]"
    RefID:int="__bind__:head->refId"
    RefIndex:int="__bind__:head->refIndex"
    DataSize:int="__bind__:head->dataSize"
    DataItemNum:int="__bind__:head->dataItemNum"
    Data:any

    def __init__(self,size) -> None:
        pass
    def IsEmpty(self)->bool:
        pass
    def Clone(self,df:"DataFrame"):
        pass

if __name__ == '__main__':
    import numpy as np
    df = DataFrame(1024)
    df_t = type(df)
    t2 = type(DataFrame)
    df.Data = np.random.rand(3,3)
    df.Format0 =10000
    uid = pyjit.generate_uid()
    df.SourceId = uid

    print(df)

File: dotnet/CantorClient/Frame.cs
using System;

namespace Cantor
{
    public static class CONST
    {
        public static int DF_FMT_NUM = 8;
        public static int METADATA_NUM = 4;
    }
    public struct UID
    {
        public UInt64 h;
        public UInt64 l;
    }
    public class DataFrameHead
    {
        public const UInt32 Tag = 0x6466726d;//'dfrm'
        public UInt32 Version = 0;

        public UInt64 Type = 0;
        public UInt64 StartTime = 0;

        public UID SourceId;
        public UID SrcAddr;
        public UID DstAddr;

        public UInt64[] Format = new UInt64[CONST.DF_FMT_NUM];
        public UInt64[] Metadata = new UInt64[CONST.METADATA_NUM];

        public UInt32 RefId = 0;
        public UInt32 RefIndex = 0;

        public UInt64 DataSize = 0;
        public UInt64 DataItemNum = 0;//bytesdata, if >=1, Key(string):Val 
    }

    public class DataFrame
    {
        public DataFrameHead? Head;
        public byte[]? Data;
        public UInt64 Context=0;
        public UInt64 Session=0;
    }
}
